llvm/lib/IR/ConstantRangeList.cpp

```diff
@@ -39,12 +39,14 @@ void ConstantRangeList::insert(const ConstantRange &NewRange) {
     return;
   assert(!NewRange.isFullSet() && "Do not support full set");
   assert(NewRange.getLower().slt(NewRange.getUpper()));
-  assert(getBitWidth() == NewRange.getBitWidth());
   // Handle common cases.
   if (empty() || Ranges.back().getUpper().slt(NewRange.getLower())) {
     Ranges.push_back(NewRange);
     return;
   }
+
+  assert(getBitWidth() == NewRange.getBitWidth());
```

Comment:
I notice we're moving a bunch of asserts around. I think that needs explaining in the commit message. However, it feels like something's wrong.

Does `insert` perhaps need to be updated so that it inserts 32-bit values depending on the range type? Would that fix us having mixed 64- and 32-bit range values? Is that even possible right now - what if we want to insert into an empty range? We can't know what bitwidth to use.

Reply:
An empty range has no bitwidth. I don't think it's worth adding an explicit field for bit width to give a bit width to the empty range

Reply:
I agree that an empty range has no bitwidth. I missed that `ConstantRangeList` only deals with the insertion of pre-existing `ConstantRange`s. I think it's okay on balance to require users to only insert/union/intersect `ConstantRange`s of matching bitwidths.

I think a good way of expressing this would be to keep the asserts where they are but have `getBitWidth` return a value that represents "none" or "empty" which always compares `==` true with bitwidths of other non-empty ranges. That's probably not worth the effort, though. What you have now is probably okay.